bfdriver/efi/efi_vmexits.cpp

```diff
@@ -0,0 +1,382 @@
+//
+// Bareflank Extended APIs
+// Copyright (C) 2018 Assured Information Security, Inc.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#include <bfvmm/vcpu/vcpu_factory.h>
+#include <bfvmm/debug/serial/serial_ns16550a.h>
+#include <eapis/vcpu/arch/intel_x64/vcpu.h>
+#include <eapis/hve/arch/intel_x64/ept/memory_map.h>
+#include <eapis/hve/arch/intel_x64/ept/helpers.h>
+
+#include <bfvmm/memory_manager/arch/x64/map_ptr.h>
+
+using namespace eapis::intel_x64;
+
+namespace efi
+{
+
+// -----------------------------------------------------------------------------
+// Handlers
+// -----------------------------------------------------------------------------
+
+static bool
+handle_cpuid(gsl::not_null<bfvmm::intel_x64::vmcs *> vmcs)
+{
+    if (vmcs->save_state()->rax == 0xBF01 || vmcs->save_state()->rax == 0xBF00) {
+        // bfvmm handles
+        return false;
+    }
+
+    auto leaf = vmcs->save_state()->rax;
+    auto ret =
+        ::x64::cpuid::get(
+            gsl::narrow_cast<::x64::cpuid::field_type>(vmcs->save_state()->rax),
+            gsl::narrow_cast<::x64::cpuid::field_type>(vmcs->save_state()->rbx),
+            gsl::narrow_cast<::x64::cpuid::field_type>(vmcs->save_state()->rcx),
+            gsl::narrow_cast<::x64::cpuid::field_type>(vmcs->save_state()->rdx)
+        );
+
+    vmcs->save_state()->rax = ret.rax;
+    vmcs->save_state()->rbx = ret.rbx;
+    vmcs->save_state()->rdx = ret.rdx;
+    if (leaf == 0x00000001) {
+        vmcs->save_state()->rcx = (ret.rcx) & (~(1UL << 26)) & (~(1UL << 27)) & (~(1UL << 5));
```

Comment:
please place bit operations in a function. Ideally, if there is a intrinsics namespace that does what you need, you should use those. Look under hypervisor/bfintrinsics/include. If not, then put the operations in a meaningful name